### PR TITLE
[#264] cxx bindings for attributes

### PR DESCRIPTION
**Pull request by elfenpiff**
Adds the c/c++ binding for attributes and integrates them into the service builder.

## Pre-Review Checklist for the PR Author

* [x] Add sensible notes for the reviewer
* [x] PR title is short, expressive and meaningful
* [x] Consider switching the PR to a draft (`Convert to draft`)
    * as draft PR, the CI will be skipped for pushes
* [x] Relevant issues are linked in the [References](#references) section
* [x] Every source code file has a copyright header with `SPDX-License-Identifier: Apache-2.0 OR MIT`
* [x] Branch follows the naming format (`iox2-123-introduce-posix-ipc-example`)
* [x] Commits messages are according to this [guideline][commit-guidelines]
    * [x] Commit messages have the issue ID (`[#123] Add posix ipc example`)
    * [x] Commit author matches [Eclipse Contributor Agreement][eca](and ECA is signed)
* [x] Tests follow the [best practice for testing][testing]
* [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
* [x] Assign PR to reviewer
* [x] All checks have passed (except `task-list-completed`)

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/doc/release-notes/iceoryx2-unreleased.md

## Checklist for the PR Reviewer

* [x] Commits are properly organized and messages are according to the guideline
* [x] Unit tests have been written for new behavior
* [x] Public API is documented
* [x] PR title describes the changes

## Post-review Checklist for the PR Author

* [x] All open points are addressed and tracked via issues

## References

Relates to #264

### CODE REVIEW

File: iceoryx2-ffi/cxx/include/iox2/attribute_verifier.hpp [hunk 1]
```diff
@@ -13,33 +13,50 @@
...
+
+  private:
+    template <ServiceType>
+    friend class ServiceBuilderEvent;
+    template <typename, typename, ServiceType>
+    friend class ServiceBuilderPublishSubscribe;
+
+    void drop();
+
+    iox2_attribute_verifier_h m_handle;
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/include/iox2/attribute_verifier.hpp:59]:
```suggestion
    iox2_attribute_verifier_h m_handle { nullptr };
```

File: iceoryx2-ffi/cxx/src/attribute_verifier.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
+// which is available at https://opensource.org/licenses/MIT.
+//
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+#include "iox2/attribute_verifier.hpp"
+
+namespace iox2 {
+AttributeVerifier::AttributeVerifier()
+    : m_handle { nullptr } {
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/src/attribute_verifier.cpp:17]: This line can be removed if the handle is initialized in the header

File: iceoryx2-ffi/cxx/tests/src/service_event_tests.cpp [hunk 3]
```diff
@@ -361,4 +361,65 @@ TYPED_TEST(ServiceEventTest, service_can_be_opened_when_there_is_a_listener) {
         ASSERT_THAT(temp_sut.has_value(), Eq(true));
     }
 }
+
+TYPED_TEST(ServiceEventTest, create_with_attributes_sets_attributes) {
+    constexpr ServiceType SERVICE_TYPE = TestFixture::TYPE;
+    constexpr uint64_t NUMBER_OF_SUBSCRIBERS = 12;
+
+    auto key = Attribute::Key("want to make your machine run faster:");
+    auto value = Attribute::Value("sudo rm -rf /");
```
Comment (elBoberido) [inline iceoryx2-ffi/cxx/tests/src/service_event_tests.cpp:370]: :grimacing:

File: iceoryx2-ffi/ffi/src/api/attribute.rs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+/// # Safety
+///
+/// * The `handle` must be a valid handle.
+#[no_mangle]
+pub unsafe extern "C" fn iox2_attribute_key_len(handle: iox2_attribute_h_ref) -> usize {
+    debug_assert!(!handle.is_null());
+
+    let attribute = (*handle).underlying_type();
+    attribute.key().len() + ZERO_TERMINATOR_LEN
+}
```
Comment (elBoberido) [inline iceoryx2-ffi/ffi/src/api/attribute.rs:47]: This is kind of counterintuitive on how, e.g. `strlen` works, which returns the length without the null terminator. Why did you choose to go this route?

File: iceoryx2-ffi/cxx/include/iox2/attribute_set.hpp [hunk 5]
```diff
@@ -13,26 +13,44 @@
...
-    auto get(const Attribute::Key& key) const -> iox::vector<Attribute::Value, IOX2_MAX_VALUES_PER_ATTRIBUTE_KEY> {
-        IOX_TODO();
-    }
-};
+    /// Returns the number of [`Attribute`]s stored inside the [`AttributeSet`].
+    auto len() const -> uint64_t;
+
+    /// Returns a [`AttributeView`] at a specific index. The number of indices is returned via
+    /// [`AttributeSetView::len()`].
+    auto at(uint64_t index) const -> AttributeView;
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx2-ffi/cxx/include/iox2/attribute_set.hpp:34]: Or we work with `begin()` and `end()` and iterators or `for_each` and `for`.

File: iceoryx2-ffi/cxx/src/attribute.cpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "iox2/attribute.hpp"
+
+namespace iox2 {
+auto AttributeView::key() const -> Attribute::Key {
+    // NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays, modernize-avoid-c-arrays) used as an uninitialized buffer
+    char buffer[Attribute::Key::capacity()];
+    iox2_attribute_key(m_handle, &buffer[0], Attribute::Key::capacity());
+    return { iox::TruncateToCapacity, &buffer[0] };
+}
```
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx2-ffi/cxx/src/attribute.cpp:21]: nice! I didn't know that this functionality existed!

File: iceoryx2-ffi/cxx/include/iox2/attribute_set.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx2-ffi/cxx/include/iox2/attribute_set.hpp:34]: Well, that's a different use case. If someone really wants to access element at index N, then it only works by checking the size before using `at` ... but that's a different philosophical question

**Review by elBoberido — APPROVED** [review]

**Pull request merged**